### DISCUSSION TITLE
neutrinordp: Allow keyboard layout information to be sent to remote.

### DISCUSSION THREAD

**Discussion by TOMATO-ONE** [neutrinolabs/xrdp#1933] [Ideas]
Hi. all

This is a discussion on Additional features for xrdp NutrinoRDP Proxy mode.
I am open to suggestions.

In `xrdp-neutrinordp.c`, the member variable of the settings structure that defines the remote keyboard layout is not set to a value, so the default value of NeutrinoRDP, `0x00`, is sent to the remote.

This means that the remote connection will not have the correct keyboard layout, except for US ASCII.
I don't know much about non-US keyboard layouts, so maybe it's not a big problem at the moment.

However, the Japanese keyboard has a special keyboard layout in the world.
When an RDP-Client with a Japanese keyboard is connected to remote Windows in xrdp RDP-Proxy mode, the US-EN keyboard layout causes problems, especially with misplaced symbols.

So far I've been working around this by changing the registry that defines the keyboard layout on the remote Windows side.
I would like to be able to deal with this on the xrdp side.

**Comment by TOMATO-ONE** [comment]
Specifically, we are considering the following measures.

As a prerequisite, the keyboard layout of the RDP Client can be referenced by the following member variables of the mod structure
```
mod->client_info.keyboard_type
mod->client_info.keyboard_subtype
mod->client_info.keylayout
```

The keyboard layout set for a Proxy connection is defined by the following member variables of the settings structure
```
<freerdp/settings.h>
settings->kbd_type
settings->kbd_subtype
settings->kbd_fn_keys
settings->kbd_layout
```

I'm thinking of bridging these two structure member variables in `xrdp-neutrinordp.c`.

More specifically, if you have `neutrinordp.allow_client_keyboardlayoyut = true` in `xrdp.ini`, then
In the `lxrdp_set_param()` function, `substitute mod->client_info.keylayout` for `settings->kbd_layout`.

https://github.com/neutrinolabs/xrdp/blob/53e2ecff4d01e6efc197bdc31d4e64225063e169/neutrinordp/xrdp-neutrinordp.c#L463-L544

**Reply by TOMATO-ONE** [reply]
I would like to enable this only when explicitly set to `true` by the admin, as in #1903.
`neutrinordp.allow_client_keyboardlayout = [true|false]`.

For most languages, we expect that simply assigning `mod->client_info.keylayout` to `settings->kbd_layout` will work, but for Japanese it may not.
You may need to mask the upper 16 bits of the keylayout, as you do in `lang.c.`
https://github.com/neutrinolabs/xrdp/blob/53e2ecff4d01e6efc197bdc31d4e64225063e169/xrdp/lang.c#L232-L249

For this reason, I'm going to add a process that looks at `mod->client_info.keyboard_type` and assigns only the lower 16 bits of `mod->client_info.keylayout` to `settings->kbd_layout` when  `7` ...  Japanese keyboard is used.

**Comment by matt335672** [comment]
It certainly looks like a good idea to me.

Since the neutrinordp module is just a proxy, the normal use-case would seem (to me) to pass the keyboard layout through if one is specified.

See also [section 2.2.1.3.2](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpbcgr/00f1da4a-ee9c-421a-852f-c19f92343d73) in [MS-RDPBCGR] which has a bit more info on this. In particular, the keyboardType can be zero, in which case the server is expected to specify the value.

**Reply by TOMATO-ONE** [reply]
@matt335672 
Thank you.

Thanks for pointing this out.
 I see that the `keyboardType` value `0x00` should be specified on the remote side, not on us-en.
I see. I understand now.

I'll explain the problem I'm having.

Assumptions
 For example, when I type `[shif] + [2]`, the us-en pc101 (ANSI 154-1988) displays `@`.
In the Japanese keyboard layout (aka. JIS X 6002-1980), `[shift]+[2]` must be typed as `"`.

It would be nice if Windows would set the keyboard layout to JIS X 6002-1980 when the `keyboardType` value is `0x00`, but that is not the case.
It will be ANSI 154-1988 compliant.

For this reason, when I type `[Shift]+[2]` on a Japanese keyboard, the `@` character is displayed.

The reason for this strange behavior is as follows.
Windows is designed to allow you to type Japanese using a standard us-en pc101 keyboard, even if it is not a Japanese keyboard.
I believe this is the default value because the ANSI 154-1988 sequence is appropriate for world standards.

I have been doing the following workaround with remote Windows.
Using a registry editor, find the following entry and enter
**_HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\Keyboard Layouts\00000411_**
Change the value of the Layout File key from `KBDJPN.DLL` to `kbd106.DLL`.

I think this was changing the default keyboard layout on the RDP server side from Japanese using ANSI 154-1988 (`KBDJPN.DLL`) to JIS X 6002-1980 (`kbd106.DLL`) when `keyboardType` is `0x00`.

In Japan, JIS X 6002-1980 is the standard Japanese keyboard layout, so it is not practical to change the Windows setting in the registry.
Let me deal with this on the xrdp side.

If the following parameter is set to true in xrdp.ini, the keyboard layout retrieved from the RDP client can be sent to the remote computer.
`neutrinordp.allow_client_keyboardlayout =[true|false]`.

**Reply by TOMATO-ONE** [reply]
This is just my imagination.
I wonder if users with US Davrok keyboards have the same problem as me?

**Reply by matt335672** [reply]
I can't add a lot to your description above I'm afraid.
I've just got UK keyboards here, which are very similar to the US ones. The " and @ characters are transposed, and the # is in a different place, but they are very similar.

**Comment by TOMATO-ONE** [comment]
I have opened Draft PR #1934 for demonstration.
I welcome suggestions for corrections to redundant code or confusing comments.

**Comment by matt335672** [comment]
For those reading this, the PR #1934 raised by @TOMATO-ONE has now been merged into devel.

**Comment by TOMATO-ONE** [comment]
Hi all
In the discussion of PR #1934, the hack about Japanese keyboard to NeutrinoRDP Proxy was excluded from PR #1934.

I was going to propose the PR again, but I'm abandoning it.

It is true that masking the upper 16 bits of the keyboardLayout is necessary when connecting to xrdp from a Japanese Windows environment with a Japanese keyboard.
I thought it was necessary to do the same for RDP Proxy connections using the NeutrinoRDP module.
However, it seems that this is not necessary.

We have implemented an option in PR #1934 to override keyboard settings when connecting to a NeutrinoRDP RDP Proxy.
I tried to use it to connect to remote Windows with various values.
The results show that there is no problem with the Japanese keyboard layout on remote connections without masking the upper 16 bits.
It seems to be properly handled by remote Microsoft Windows.

I made a false assumption in PR #1934. Sorry about that.

The original purpose has already been achieved.

Based on the findings of this discussion, I've started a proposal in #1950 for the ability to override keyboard information in xrdp itself.

Please continue to join in the discussion.
